### PR TITLE
config: add `region-max-size ` property

### LINKED ISSUE #4459: sync tikv config
## Enhancement Task
Config supports `region-max-size` property. This value should only be set by file not by api.

### PR DESCRIPTION
**Pull request by bufferflies**
Signed-off-by: bufferflies <1045931706@qq.com>

### What problem does this PR solve?

Issue Number: Close #4459

### What is changed and how it works?
`config.PersistOptions` should save the initial config and not update it .

```commit-message
```

### Check List

Tests

- Unit test

Code changes

Side effects

Related changes

### Release note

```release-note
config adds`region-max-size ` property 
```

### CODE REVIEW

File: server/config/persist_options.go [hunk 1]
```diff
@@ -50,11 +50,12 @@ type PersistOptions struct {
 	replicationMode atomic.Value
 	labelProperty   atomic.Value
 	clusterVersion  unsafe.Pointer
+	cfg             *Config
```
Comment (rleungx) [inline server/config/persist_options.go:53]: How about adding an interface function into Cluster?

**Comment by rleungx** [comment]
Please add a unit test to check if changing the parameter in the config file works.

File: server/config/persist_options.go (hunk 1 above)
Comment (bufferflies) [reply server/config/persist_options.go:53]: The most values of the  persistOptions is transferred by config and I think the cluster should export one method to manage all config.

Comment (rleungx) [reply server/config/persist_options.go:53]: Those configs in persist option can be changed dynamically, I prefer not to put `cfg` inside it.

**Comment by nolouch** [comment]
How this config influence the TiKV?

File: conf/config.toml [hunk 2]
```diff
@@ -23,7 +23,7 @@
 
 ## Join to an existing cluster. The value should be cluster's ${advertise-client-urls}
 # join = ""
-
+# max-region-size = 96
```
Comment (disksing) [inline conf/config.toml:26]: insert an empty line?

File: pkg/mock/mockcluster/mockcluster.go [hunk 3]
```diff
@@ -51,18 +51,20 @@ type Cluster struct {
...
-		BasicCluster:   core.NewBasicCluster(),
-		IDAllocator:    mockid.NewIDAllocator(),
-		HotStat:        statistics.NewHotStat(ctx),
-		PersistOptions: opts,
-		suspectRegions: map[uint64]struct{}{},
+		BasicCluster:    core.NewBasicCluster(),
+		IDAllocator:     mockid.NewIDAllocator(),
+		HotStat:         statistics.NewHotStat(ctx),
+		PersistOptions:  opts,
+		ImmutableConfig: config.NewTestImmutableOptions(),
```
Comment (disksing) [inline pkg/mock/mockcluster/mockcluster.go:66]: Does `immutable` means that it cannot be updated by pd-ctl?

File: pkg/mock/mockcluster/mockcluster.go [hunk 4]
```diff
@@ -78,6 +80,11 @@ func (mc *Cluster) GetOpts() *config.PersistOptions {
 	return mc.PersistOptions
 }
 
+// GetConfig returns the cluster immutable configuration.
+func (mc *Cluster) GetConfig() *config.ImmutableConfig {
```
Comment (rleungx) [inline pkg/mock/mockcluster/mockcluster.go:84]: Prefer to change the name.

File: server/config/immutable_options.go [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+// ImmutableConfig is a readonly config.
+type ImmutableConfig struct {
+	maxRegionSize uint64
+	leaderLease   int64
```
Comment (rleungx) [inline server/config/immutable_options.go:20]: Why only leaderLease?

File: server/config/immutable_options.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+// GetMaxRegionSize returns the max region size.
+func (c *ImmutableConfig) GetMaxRegionSize() uint64 {
+	return c.maxRegionSize
+}
+
+// ImmutableConfigCreateOption used to create ImmutableConfig.
+type ImmutableConfigCreateOption func(config *ImmutableConfig)
+
+// WithMaxRegionSize set max region size.
+func WithMaxRegionSize(maxRegionSize uint64) ImmutableConfigCreateOption {
```
Comment (rleungx) [inline server/config/immutable_options.go:44]: As it is an immutable config, perhaps we don't need this function.

File: server/schedule/operator/builder.go [hunk 7]
```diff
@@ -33,6 +33,7 @@ import (
 type ClusterInformer interface {
 	GetBasicCluster() *core.BasicCluster
 	GetOpts() *config.PersistOptions
+	GetConfig() *config.ImmutableConfig
```
Comment (rleungx) [inline server/schedule/operator/builder.go:36]: How about moving it to `Cluster`?

**Comment by bufferflies** [comment]
> How this config influence the TiKV?

the configuration will not influence the TiKVs, maybe can consider this pr:#4308

File: pkg/mock/mockcluster/mockcluster.go (hunk 3 above)
Comment (bufferflies) [reply pkg/mock/mockcluster/mockcluster.go:66]: yes, pd will reload config after restarting.

File: conf/config.toml (hunk 2 above)
Comment (bufferflies) [reply conf/config.toml:26]: fix

File: server/schedule/operator/builder.go (hunk 7 above)
Comment (bufferflies) [reply server/schedule/operator/builder.go:36]: this immutable config can be used by some checks or schedulers.

Comment (bufferflies) [reply server/schedule/operator/builder.go]: fix

File: pkg/mock/mockcluster/mockcluster.go (hunk 4 above)
Comment (bufferflies) [reply pkg/mock/mockcluster/mockcluster.go]: fix to GetImmutableCfg

File: server/config/immutable_options.go (hunk 5 above)
Comment (bufferflies) [reply server/config/immutable_options.go]: `leaderLease` will be  remove and add `max-split-size`.

**Comment by bufferflies** [comment]
It is not graceful to add some tikv config into the config file, so I will use call back tikv to get this config.

**Pull request closed without merging**